alternator/ttl.cc

```diff
@@ -261,6 +263,187 @@ static future<> expire_item(service::storage_proxy& proxy,
         qs.get_trace_state(), qs.get_permit());
 }
 
+// A class for iterating over all the token ranges *owned* by this shard.
+// We consider a token *owned* by this shard if:
+// 1. This node is a replica for this token.
+// 2. Moreover, this node is the *primary* replica of the token (i.e., the
+//    first replica in the ring).
+// 3. In this node, this shard is responsible for this token.
+// We will use this definition of which shard in the cluster owns which tokens
+// to split the expiration scanner's work between all the shards of the
+// system.
+// When some of the nodes in the system are down, nobody will be expiring
+// tokens owned by them, so we will need to separately detect when this
+// happens and let living nodes take over the work that a dead node cannot
+// do.
+// FIXME: need to decide how to choose primary ranges in multi-DC setup!
+// We could call get_primary_ranges_within_dc() below instead of get_primary_ranges().
+class token_ranges_owned_by_this_shard {
+    schema_ptr _s;
+    // _token_ranges will contain a list of token ranges owned by this node.
+    // We'll further need to split each such range to the pieces owned by
+    // the current shard, using _intersecter.
+    dht::token_range_vector _token_ranges;
+    dht::token_range_vector::const_iterator _range_it;
+    std::optional<dht::selective_token_range_sharder> _intersecter;
+public:
+    // FIXME: add a constructor to start iteration in the middle, to continue
+    // a scan after reboot.
+    token_ranges_owned_by_this_shard(database& db, schema_ptr s)
+        :  _s(s)
+        , _token_ranges(db.find_keyspace(s->ks_name()).
+            get_effective_replication_map()->get_primary_ranges(
+            utils::fb_utilities::get_broadcast_address()))
+        , _range_it(_token_ranges.cbegin())
+    {}
+
+    // Return the next token_range owned by this shard, or nullopt when the
+    // iteration ends.
+    std::optional<dht::token_range> next() {
+        if (_intersecter) {
+            std::optional<dht::token_range> ret = _intersecter->next();
+            if (ret) {
+                return ret;
+            }
+            // done with this range, go to next one
+            ++_range_it;
+            _intersecter = std::nullopt;
+        }
+        if (_range_it == _token_ranges.cend()) {
+            return std::nullopt;
+        }
+        _intersecter.emplace(_s->get_sharder(), *_range_it, this_shard_id());
+        // _intersecter.next() can't be empty, because we can assume that
+        // every vnode must intersect with every shard (we don't have vnodes
+        // that aren't serviced by some of the shards). If we didn't make this
```

Comment:
Why do you assume this?

Reply:
Trivial counterexample: a vnode (7, 8].

Reply:
Oh :-(
I was assuming that vnodes are roughly of the same *large* size. I forgot they are completely random, so can be tiny.
Of course, it's bad when it's tiny, but I can't rule it out :-(
So I need to go back to the loop I was hoping to avoid. Oh, well.

Reply:
Will be fixed in the next version.